Added vendored package blog post (#269)
conan-io/conan-io.github.io: _posts/2024-06-24-Introducing-vendoring-packages.markdown

    $ mkdir lib_a && cd lib_a
    $ conan new cmake_lib -d name=lib_a -d version=1.0
    $ conan create .
    ```

2. Create a package depending on the previous library, which will be the one we’ll use to vendor its dependencies

    ```sh
    $ cd .. && mkdir sdk && cd sdk
    $ conan new cmake_lib -d name=sdk -d version=1.0 -d requires="lib_a/1.0"
    $ conan create .
    ```

3. Create a consumer application depending on the sdk library:

    ```sh
    $ cd .. && mkdir app && cd app
    $ conan new cmake_exe -d name=app -d version=1.0 -d requires="sdk/1.0" 
    ```

4. Install the created application 

    ```sh
    $ conan install . --build=missing

    ======== Computing dependency graph ========
    ...
    Requirements
        lib_a/1.0#ab64452c42599a3dc0ee6a0dc90bbd90 - Cache
        sdk/1.0#1cb781c232f63845b7943764d8a084ed - Cache

    ======== Computing necessary packages ========
    Requirements
        lib_a/1.0#ab64452c42599a3dc0ee6a0dc90bbd90:39f48664f195e0847f59889d8a4cdfc6bca84bf1#e34a89988cafb2bf67f6adf40b06f442 - Cache
        sdk/1.0#1cb781c232f63845b7943764d8a084ed:12ffb661ea06cee312194b5f6acd48e8236b8ed8#9127cf762dfd1a1f505ecd1d3ac056b9 - Cache
    ```

>     Transitive dependencies are required as expected.

Reviewer: ```suggestion
    We can see that transitive dependencies are required as expected.
```

        # Sources are located in the same place as this recipe, copy them to the recipe
        exports_sources = "CMakeLists.txt", "src/*", "include/*"

        def config_options(self):
            if self.settings.os == "Windows":
                self.options.rm_safe("fPIC")

        def layout(self):
            cmake_layout(self)

        def requirements(self):
            self.requires("lib_a/1.0")

        def generate(self):
            deps = CMakeDeps(self)
            deps.generate()
            tc = CMakeToolchain(self)
            tc.generate()

        def build(self):
            cmake = CMake(self)
            cmake.configure()
            cmake.build()

        def package(self):
            # Repackage static dependencies inside the package
            copy(self, "*.a", src=self.dependencies["lib_a"].cpp_info.libdir, dst=os.path.join(self.package_folder, self.cpp_info.libdir))
            cmake = CMake(self)
            cmake.install()

        def package_info(self):
            self.cpp_info.libs = ["sdk", "lib_a"]

    ```

7. Apply the changes and create the SDK package:

    ```sh
    $ cd ../sdk
    $ conan create .
    ```

> 8. Finally, let’s reinstall the application again.

Reviewer: ```suggestion
8. Finally, let's reinstall the application again.
```

    ```sh
    $ cd ../app 
    $ conan install . --build=missing

    ======== Computing dependency graph ========
    Graph root
    ...
    Requirements
        sdk/1.0#f2fd2a793849725303073d37b15042b2 - Cache

    ======== Computing necessary packages ========
    Requirements
        sdk/1.0#f2fd2a793849725303073d37b15042b2:5d605f63db975c8c6004cc0a0b5c99c99dce6cc3#92c538ec767c2ff02a2fddf6b4106d02 - Cache
    ```

>     As it can be seen, while computing the dependency graph, conan does not

Reviewer: ```suggestion
    As it can be seen, while computing the dependency graph, Conan does not
```

    retrieve either the recipe of ``lib_a/1.0`` nor the binaries. 

9. We could try to even remove ``lib_a`` from our local cache and install again the application:

    ```sh
    $ conan remove "lib_a*" -c
    Found 1 pkg/version recipes matching lib_a* in local cache
    Remove summary:
    Local Cache
      lib_a/1.0#ab64452c42599a3dc0ee6a0dc90bbd90: Removed recipe and all binaries
    ```

    ```sh
    $ conan install .
    ...
    Requirements
        sdk/1.0#f2fd2a793849725303073d37b15042b2 - Cache
    ...
    Install finished successfully
    ```

>     It works!!! Here is where vendoring feature shines. Package creators can

Reviewer: ```suggestion
    It works!!! Here is where the vendoring feature shines. Package creators can
```

Now it's referred to as "vendoring" feature when on top it was vendor, goes to show it reads better :)

>     distribute their packages without needing to distribute their private

Reviewer: ```suggestion
    distribute their packages without the need to distribute their private
```

    dependencies.

> 10. Generate graph info of the application using vendored SDK

Reviewer: ```suggestion
10. Generate the graph info of the application using the vendored SDK
```

- **Efficiency**: Reducing download and build times by not fetching unnecessary components.
- **Control**: Allowing software vendors to manage how their packages are used and distributed without exposing internal details.

### Forcing the build of a vendoring package

We have seen how a vendor can be used from a consumer directly downloading
binaries without needing to download any dependency data.

But what happens if we want to compile the internal vendored dependencies?

1. As we have previously deleted ``lib_a`` package, we should re create it:

    ```sh 
    $ cd ../lib_a && conan create .
    ```
    Now, let’s force the build of our previous SDK example:

    ```sh
    $ cd ../app
    $ conan install . --build="sdk/1.0"
    ...
    ======== Computing necessary packages ========
    sdk/1.0: Forced build from source
    Requirements
        sdk/1.0#f2fd2a793849725303073d37b15042b2:5d605f63db975c8c6004cc0a0b5c99c99dce6cc3 - Invalid
    ERROR: There are invalid packages:
    sdk/1.0: Invalid: The package 'sdk/1.0' is a vendoring one, needs to be built from source, but it didn't enable 'tools.graph:vendor=build' to compute its dependencies
    ```

>     Trying to build a vendor package will fail by default unless setting the ``tools.graph:vendor`` configuration to “build”.

Reviewer: ```suggestion
    Trying to build a vendor package will fail by default we set the ``tools.graph:vendor`` configuration to “build”.
```

> 2. Once the ``vendor`` configuration is enabled, the user must have access to packaged dependencies as if it were a normal package.

Reviewer: ```suggestion
2. Once the ``vendor`` configuration is enabled, the user must have access to the packaged dependencies as if it were a normal package.
```

    ...
    ======== Computing necessary packages ========
    sdk/1.0: Forced build from source
    Requirements
        lib_a/1.0#ab64452c42599a3dc0ee6a0dc90bbd90:39f48664f195e0847f59889d8a4cdfc6bca84bf1#5a8bfa1c980c2008e7e24996a4b48477 - Cache
        sdk/1.0#f2fd2a793849725303073d37b15042b2:5d605f63db975c8c6004cc0a0b5c99c99dce6cc3 - Build

    ======== Installing packages ========
    lib_a/1.0: Already installed! (1 of 2)
    ...
    Install finished successfully
    ```

    When forcing the compilation of a vendored dependency, the graph expands again,
    revealing the encapsulated dependencies necessary for the build.

3. Let's generate the graph again to see how the expansion works:

    ```sh
    $ conan graph info . --build="sdk/1.0" -c tools.graph:vendor=build --format=html > vendored-expanded-graph.html
    ```

    <div style="display: flex; flex-direction: column; align-items: center; margin: 0 10px;">
        <img src="{{ site.baseurl }}/assets/post_images/2024-06-24/vendored-expanded-graph.png" alt="Vendored expanded graph" style="width: 25%;" />
        <span>Vendored expanded graph</span>
    </div>
    <br/>

>     **Note**: Red dashed border means the package is a vendor and yellow background means the package has been forced to be built

Reviewer: ```suggestion
    **Note**: Red dashed borders means the package is a vendor and yellow background means the package has been forced to be built
```

> 4. To verify a vendored package does not need to have their transitive

Reviewer: ```suggestion
4. To verify that a vendored package does not need to have their transitive
```

   dependencies accessible unless forced to build, we can try to remove our ``lib_a`` package from our local cache and install again:

    ```sh
    $ conan remove "lib_a*" -c
    Found 1 pkg/version recipes matching lib_a* in local cache
    Remove summary:
    Local Cache
      lib_a/1.0#ab64452c42599a3dc0ee6a0dc90bbd90: Removed recipe and all binaries
    ```

    ```sh
    $ conan install . --build="sdk/1.0" -c tools.graph:vendor=build
    ...
    ======== Computing dependency graph ========
    lib_a/1.0: Not found in local cache, looking in remotes...
    lib_a/1.0: Checking remote: conancenter
    ...
    Requirements
        sdk/1.0#f2fd2a793849725303073d37b15042b2 - Cache
    ERROR: Package 'lib_a/1.0' not resolved: Unable to find 'lib_a/1.0' in remotes.
    ```

>     As expected, an error package not resolved is thrown when forcing building the vendored package if we do not have access to its dependencies.

Reviewer: ```suggestion
    As expected, an "Error: Package 'lib_a/1.0' not resolved" is raised when forcing the build of the vendored package if we do not have access to its dependencies.
```

### Proxy vendor for Extra Privacy

For vendors that want to further obscure their dependency names and versions, a
proxy vendored package can be created. This proxy package should include all
internal/private dependencies and should be made inaccessible to clients. 

The primary vendored recipe depends on this proxy package, and by marking both
packages as vendored, the actual dependencies and their versions will be hidden
from the end-user.

### Impact on Package ID Calculation

To support the vendor feature, we’ve adjusted how the ``package_id`` is calculated
for vendored packages. Since dependencies and their versions are invisible to
the end-user, they are excluded from the ``package_id`` calculation. Only the
recipe revision will change with recipe content changes, but the ``package_id``
will remain the same. This ensures that any kind of change to internal
dependencies (updating, adding or removing) will not alter the ``package_id``,
saving time and effort in environments where consistent package IDs are
crucial.

> It goes without saying that the provider in charge of packaging the package

Reviewer: ```suggestion
It goes without saying that the provider of packaging the package
```

should be in charge of updating the version of the package when any of its

> internal dependencies change version in order to update the ``package_id`` and let

Reviewer: Updating the version won't update the package_id!

Author: Yep, sorry, fixed it!
